### PR TITLE
Fixed -- 28785 Made makemigrations warn if migrations are missing

### PR DESCRIPTION
**Pull request by manav014**
[ticket-28785](https://code.djangoproject.com/ticket/28785)

Supersedes [#9339  ](https://github.com/django/django/pull/9339)

### CODE REVIEW

**Comment by olivierdalang** [comment]
Thanks for this !

IMO, this check should not only run for makemigrations, but also for other commands  (e.g. runserver, createsuperuser, etc.)

For unapplied migrations, [management/base.py](https://github.com/django/django/blob/ca9872905559026af82000e46cde6f7dedc897b6/django/core/management/base.py#L479-L504) produces this warning : 
> You have 2 unapplied migration(s). Your project may not work properly until you apply the migrations for app(s): core. Run 'python manage.py migrate' to apply them.

Both cases are quite similar, so having the same type of warning for both would make sense. Can we have something like :
>You have 2 applied migration(s) that are missing the original migration file for app(s): core. Ensure you did not delete or renamed migrations files, and/or that you are working on the correct branch. Your project may not work properly until you fix this situation.

**Comment by jacobtylerwalls** [comment]
> IMO, this check should not only run for makemigrations, but also for other commands (e.g. runserver, createsuperuser, etc.)

This sounds vague. I would rather the author stick to the scope of the ticket as accepted. Doing so helps both authors and reviewers.

**Comment by olivierdalang** [comment]
I don't know what's vague in my comment (maybe the wording, I'm not a native English speaker, sorry). Management commands even have a `requires_migrations_checks` property [documented like this](https://docs.djangoproject.com/en/3.2/howto/custom-management-commands/#django.core.management.BaseCommand.requires_migrations_checks) :

> A boolean; if True, the command prints a warning if the set of migrations on disk don’t match the migrations in the database. A warning doesn’t prevent the command from executing. Default value is False.

This exactly matches the scope of what this PR is about, doesn't it ?

**Comment by jacobtylerwalls** [comment]
Ah, thanks for clarifying. Your wording was fine. 👍🏻   It just took a second read through the ticket for me to see that the makemigrations command was not singled out until after the ticket was originally accepted, my apologies.

**Comment by jacobtylerwalls** [comment]
> Management commands even have a requires_migrations_checks property documented like this :

Thanks, excellent find. In case helpful for others, here's the status quo for that property on various commands:

- `BaseCommand`: False
- `changepassword`: True
- `createsuperuser`: True
- `runserver`: categorically runs the unapplied migrations check regardless of the class attribute because that check ran as part of `runserver` before the introduction of the class attribute in 50931dfa.

@manav014 with that in mind, do you have an opinion about how to proceed here?

I would need to think more myself. The first comment on the ticket cautions this might pester folks who turn squashed migrations into normal ones, so I'm hesitant to patch this into `runserver` categorically. (Or, unlike the unapplied migrations check have this check observe the class attribute?)

Another thing is that we have ticket-26760 for dropping migrations from the db when they don't exist on disk, and some discussion about whether to execute that automatically vs. manually. That would make this situation less frequent, no? 

Thanks for your input @olivierdalang; good find, and once again, apologies if I sounded out of breath earlier.

**Comment by olivierdalang** [comment]
@jacobtylerwalls Are you referring to this ?
 
https://github.com/django/django/blob/bd9cfbb0135aeb8528716da8fa4d0d090d0f9558/docs/topics/migrations.txt#L816-L822

That's indeed a case where the warning would show. But I'm not sure the docs give a good advice there (is it mandatory to remove `replaces`, or can it list deleted migrations ?).

As for this PR, maybe it could add a line in that part of the docs warning that doing that process would lead to warnings displaying at users. And the warning itself could mention the case of squashed migrations and suggest to remove the rows in the migration tables if that's the case. It's a bit hard to tell how frequently this is being used.

All in all, I think it's a deeper issue in the squashing process (alongside other issue with it, see https://groups.google.com/g/django-developers/c/xpeFRpMTBZw/m/lDq78EedAwAJ), and I'd see this PR as beneficial anyways as it could actually help highlight some existing potential issues, including the one mentioned in ticket 26760.

**Comment by manav014** [comment]
> @manav014 with that in mind, do you have an opinion about how to proceed here?

@jacobtylerwalls I think I need to dig more into this. But as of now, What I have in mind is that we shall use `requires_migrations_checks` to show the warning that means if it is set to true we shall display the warning like in `changepassword` and `createsuperuser` commands this is set to `true`.

**Comment by jacobtylerwalls** [comment]
> Are you referring to this ?

Yep, you found the right bit in the docs re: "transition the squashed migration to a normal migration".

> That's indeed a case where the warning would show. But I'm not sure the docs give a good advice there (is it mandatory to remove replaces, or can it list deleted migrations ?).

Recent triage decisions have stuck to this literal advice that it's mandatory, see ticket-32205.

**Comment by jacobtylerwalls** [comment]
> But as of now, What I have in mind is that we shall use requires_migrations_checks to show the warning that means if it is set to true we shall display the warning like in changepassword and createsuperuser commands this is set to true.

Nice, that sounds like a good start. Then we can get a decision on `runserver` after that. Thanks.

File: django/db/migrations/loader.py [hunk 1]
```diff
@@ -285,6 +285,24 @@ def build_graph(self):
...
+        for migration in applied:
+            if migration not in self.graph.nodes:
+                deleted_migration_al.add(migration[0])
+        if deleted_migration_al:
+            if not app_labels:
+                raise InconsistentMigrationFileHistory(
+                    ', '.join(repr(app) for app in sorted(deleted_migration_al)))
+            elif any(apps in app_labels for apps in deleted_migration_al):
+                raise InconsistentMigrationFileHistory(
+                    ', '.join(repr(app) for app in sorted(deleted_migration_al) if app in app_labels))
```
Comment (felixxm) [inline django/db/migrations/loader.py:304]: I would prefer the solution from #9339 where we list all missing files not only all labels.

File: django/db/migrations/loader.py [hunk 2]
```diff
@@ -285,6 +285,24 @@ def build_graph(self):
             raise
         self.graph.ensure_not_cyclic()
 
+    def check_consistent_migration_files(self, connection, app_labels):
```
Comment (felixxm) [inline django/db/migrations/loader.py:288]: Do we need `app_labels`? IMO it should be fine to check all apps like in `check_consistent_history()`.

**Review by felixxm — COMMENTED** [review]
@manav014 Thanks for this patch :+1:

I'm afraid that we cannot move it forward without solving ticket-26760, because it will raise warnings for squashed migrations which were removed.

File: django/db/migrations/loader.py (hunk 1 above)
Comment (manav014) [reply django/db/migrations/loader.py:304]: I will commit the desired changes.

File: django/db/migrations/loader.py (hunk 2 above)
Comment (manav014) [reply django/db/migrations/loader.py:288]: I will commit the desired changes.

**Comment by manav014** [comment]
@felixxm Thank you for your feedback. 
I will make all the commented changes and also I totally agree with the point that we can not move it forward without solving [ticket-26760](https://code.djangoproject.com/ticket/26760). Till then let's wait for the [ticket-26760](https://code.djangoproject.com/ticket/26760) to be closed. I will try to assign the ticket-26760 to myself as soon as I will get some time.

**Comment by jacobtylerwalls** [comment]
Hi @manav014 just a note to say ticket-26760 was recently fixed with a feature to call `migrate --prune` to remove references to deleted migrations, so maybe in the case that was blocking you here with deleted squashed migrations you can advise the user to do that, or something similar. Hope that helps.

**Comment by manav014** [comment]
Thanks, @jacobtylerwalls . I will add a warning message to warn the user about `migrate --prune` .

**Comment by felixxm** [comment]
Closing due to inactivity.

**Pull request closed without merging**